File: Plugins/Lookups/DNSLookup.py
from Plugins import Plugin

import re
import socket

class DNSLookup(Plugin.Plugin):
  def __init__(self, dns: str = None, name: str = 'DNSLookup'):
    super().__init__(name)
    self._dns = dns

  def _performLookup(self, dns) -> dict:
    dns = re.sub("(http|https)://", "", dns)
    try:
      dns_resolution = socket.gethostbyname(dns)
    except:
      dns_resolution = "No IP association found."

    return {
      f"{dns}": {
        "dns_resolution": dns_resolution
      }
    }

  def run(self):
    print("\n -------------------------------- ")
    print("        D N S  L O O K U P        ")
    print(" -------------------------------- ")
    if self._dns == None:
      self._dns = input('Enter DNS value: ').strip()

    self._dns = re.sub("(http|https)://", "", self._dns)
    lookup = self._performLookup(self._dns)
    print(f"\nLookup returned: {lookup[self._dns]['dns_resolution']}")

File: src/plugins/standard_lookups/dns_lookup.rs
use super::Plugin;
use std::net::{IpAddr, ToSocketAddrs};
use std::collections::HashMap;

use crate::utils::get_input;

pub struct DNSLookup {}

impl Plugin for DNSLookup {
  const NAME: &'static str = "DNS Lookup";

  fn new() -> Self {
    Self {}
  }

  fn perform_lookup(&self, value: &str) -> Result<serde_json::Value, String> {
    let dns = value.trim_start_matches("http://").trim_start_matches("https://").trim();
    let default_port = 80;

    let dns_with_port = format!("{}:{}", dns, default_port);

    match dns_with_port.to_socket_addrs() {
      Ok(socket_addrs) => {
        let mut dns_resolution: Vec<String> = Vec::new();
        for socket_addr in socket_addrs {
          if let IpAddr::V4(ipv4) = socket_addr.ip() {
            dns_resolution.push(ipv4.to_string());
          } else if let IpAddr::V6(ipv6) = socket_addr.ip() {
            dns_resolution.push(ipv6.to_string());
          }
        }

        Ok(serde_json::json!({
          "status": "complete",
          "dns_resolution": dns_resolution
        }))
      }
      Err(error) => {
        println!("{}", error);
        Err(String::from("DNS resolution failed"))
      }
    }
  }

  fn run(&self) {
    println!("{}", Self::NAME);

    let usr_in = get_input("Enter domain to resolve: ");
    match self.perform_lookup(&usr_in.trim()) {
      Ok(result) => println!("\n{}", result),
      Err(error) => eprintln!("Error: {}", error)
    }
  }
}